### DISCUSSION TITLE
Dialog content rerenders and StrictMode

### DISCUSSION THREAD

**Discussion by Isaacmaamouche** [ariakit/ariakit#3414] [Questions] [answered]
Hello, 

In StrictMode, I've notice that childrens of a `Ariakit.Dialog` are rendered not twice but four times. 
I added a rerender counter and console.logged it, it doesn't go from 0 to 3, but from 0 to 1 twice : 0 1 0 1 
![Screenshot 2024-01-31 at 17 28 45](https://github.com/ariakit/ariakit/assets/85453899/73d0218b-0e1a-4c44-9ec3-bc313a6ba399)

I have reproduced the behaviour using Ariakit latest version in this [codeSandBox](https://codesandbox.io/p/devbox/welcome-ui-8y3njt?file=%2Fsrc%2FApp.tsx%3A9%2C38&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cloo2duro000a3b6hm92ailtc%2522%252C%2522sizes%2522%253A%255B61.51943462897526%252C38.48056537102474%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cloo2duro00033b6hagei4lmg%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cloo2duro00073b6hhkmy4v03%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cloo2duro00093b6huoiu8xax%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cloo2duro00033b6hagei4lmg%2522%253A%257B%2522id%2522%253A%2522cloo2duro00033b6hagei4lmg%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloo2durn00023b6hoiqxbzrh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522id%2522%253A%2522cloo2fkng00fx3b6h4c549uvs%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A9%252C%2522startColumn%2522%253A38%252C%2522endLineNumber%2522%253A9%252C%2522endColumn%2522%253A38%257D%255D%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Fstyled.d.ts%2522%252C%2522id%2522%253A%2522cloo3g63u012r3b6i38blrbl8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522cls1z014m017q3b6ga8ehxum8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cloo2fkng00fx3b6h4c549uvs%2522%257D%252C%2522cloo2duro00093b6huoiu8xax%2522%253A%257B%2522id%2522%253A%2522cloo2duro00093b6huoiu8xax%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloo2duro00083b6hwiroai48%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cloo2duro00083b6hwiroai48%2522%257D%252C%2522cloo2duro00073b6hhkmy4v03%2522%253A%257B%2522id%2522%253A%2522cloo2duro00073b6hhkmy4v03%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloo2duro00043b6h6jm3xbvq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522cloo2euw700ak3b6h87jf7y6c%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cls1yihir001adzgq9bbc6uw4%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cloo2euw700ak3b6h87jf7y6c%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
You can move the `JustADiv` component in and out the `Ariakit.Dialog` to witness the difference in terms or rerenders. 

Is it an expected behaviour ?
Would this be considered a bug ? 

Best regards 👋

**Comment by diegohaz** [comment, marked as answer]
Yes, that's expected. It's because the dialog is rendered in a portal. You can adjust it by setting `<Dialog portal={false}>` to see the difference.

**Reply by Isaacmaamouche** [reply]
All right, thank you the answer !
